kafka-client/src/main/java/dev/responsive/kafka/internal/stores/CommitBuffer.java

```diff
@@ -466,27 +468,19 @@ private void doFlush(final long consumedOffset, final int batchSize) {
         buffer.getReader().size(),
         batchSize,
         consumedOffset,
-        writerFactory
+             writeBatcher
     );
 
-    final WriterFactory<K, P>.PendingFlush pendingFlush = writerFactory.beginNewFlush();
-    for (final Result<K> result : buffer.getReader().values()) {
-      final RemoteWriter<K, P> writer = pendingFlush.writerForKey(result.key);
-
-      if (result.isTombstone) {
-        writer.delete(result.key);
-      } else if (keySpec.retain(result.key)) {
-        writer.insert(result.key, result.value, result.timestamp);
-      }
-    }
+    final RemoteWriteResult<P> flushResult = writeBatcher.flushWriteBatch(
+        buffer.getReader().values(), consumedOffset);
 
-    final var flushResult = writerFactory.commitPendingFlush(pendingFlush, consumedOffset);
-    // TODO: we could/probably should move the result checking and handling inside the
-    //  #commitPendingFlush method, since we have to go back to the WriterFactory for the
-    //  specific error message anyways. It might also be nice to extract the exception
-    //  supplier so that different tables and/or writers can throw more specific errors
     if (!flushResult.wasApplied()) {
-      throwFlushException(flushResult, consumedOffset);
+      flushErrorsSensor.record();
+
+      final String errorMsg = writeBatcher.failedFlushMsg(flushResult, consumedOffset);
```

Comment:
(suggestion - personal preference) I personally prefer creating error messages inline so that it's easy to find the place in the code based on the error message with a ctrl+F, I'd rather do that than abstract it away even if there's some code duplication (same with below)